### ISSUE TITLE
[Meta] New Backend System | Release 1.0

### ISSUE THREAD

**Issue by camilaibs** [backstage/backstage#24493]
## Table of Content

  - [Glossary](#glossary)
  - [Context](#context)
  - [Deadline](#deadline)
  - [Monitoring](#monitoring)
  - [Contributing](#contributing)
  - [Scope](#scope)
  - [Nice to Have](#nice-to-have)
  - [Out-of-Scope](#out-of-scope)

## Glossary <a href="#glossary" id="glossary"/>

<dl>
  <dt>NBS</dt>
  <dd>Abbreviation of "New backend System"</dd>
</dl>

## Context <a href="#context" id="context"/>

It is the stable release of the new backend system along with a deprecation of the old system and a huge push towards a stable plugin ecosystem built on top of the new backend system. The stable release is vital to ensure that adopters have confidence in the new system.

## Deadline <a href="#deadline" id="deadline"/>

We **estimate** that preparations for the stable version will take 2-3 months, with work starting this month, May 2024.
The stable version will be released in August-September releases, and the legacy system will be deprecated and no longer updated after that.
Next, we intend to completely remove the legacy system by the end of December 2024, although we may be able to remove it sooner if we move forward and complete the previous steps of preparation and deprecation faster.

## Monitoring <a href="#monitoring" id="monitoring"/>

This issue is mainly created for tracking the work to be done prior to the release, see a comprehensive list of tasks and their respective status in the [Scope](#scope) session.

## Contributing <a href="#contributing" id="contributing"/>

We are always open to contributions 🙏🏻 , so please comment here if you would like to work on some of the scoped tasks.

## Scope <a href="#scope" id="scope"/>

Excluding this milestone dependencies, it is primarily cleanup, review of APIs and documentation. There is also some amount of work related to ensuring the deprecation of the `@backstage/backend-common`, as well as any remaining plugin migrations of core features.

> [!Note]
Let us know by commenting in this issue if there's anything you would like us to fix or add before the release 🙌🏻 .

### Documentation

Documentation pages that need to be updated are tracked in the [external issue](https://github.com/backstage/backstage/issues/24195) below:

```[tasklist]
### Tasks
- [ ] https://github.com/backstage/backstage/issues/24195
```

### Framework

Here are tasks enumerated after a first review of the framework packages API:

#### Defaults and Common

```[tasklist]
### Tasks
- [ ] https://github.com/backstage/backstage/issues/24803
- [ ] https://github.com/backstage/backstage/issues/24851
- [ ] https://github.com/backstage/backstage/issues/24606
- [ ] https://github.com/backstage/backstage/issues/24736
- [ ] https://github.com/backstage/backstage/issues/24535
- [ ] https://github.com/backstage/backstage/issues/24542
- [ ] https://github.com/backstage/backstage/issues/24541
- [ ] https://github.com/backstage/backstage/issues/24604
- [ ] https://github.com/backstage/backstage/issues/25148
- [ ] https://github.com/backstage/backstage/issues/24549
- [ ] https://github.com/backstage/backstage/issues/24548
```

#### App API

```[tasklist]
### Tasks
- [ ] https://github.com/backstage/backstage/issues/25270
- [ ] https://github.com/backstage/backstage/issues/25271
- [ ] https://github.com/backstage/backstage/issues/25272
- [ ] https://github.com/backstage/backstage/issues/25273
```

#### Plugin API

```[tasklist]
### Core
- [ ] https://github.com/backstage/backstage/issues/24854
- [ ] https://github.com/backstage/backstage/issues/24855
- [ ] https://github.com/backstage/backstage/issues/24910
- [ ] https://github.com/backstage/backstage/issues/25121
- [ ] https://github.com/backstage/backstage/issues/24920
- [x] Evaluate if the `URLReader` is really a core service? Also reconsider if the `Schedule` should continue in the core services list as well
```

```[tasklist]
#### Service Modules
- [ ] https://github.com/backstage/backstage/issues/24923
- [ ] https://github.com/backstage/backstage/issues/25558
- [ ] https://github.com/backstage/backstage/issues/25561
- [ ] https://github.com/backstage/backstage/issues/24547
- [x] Convert the `UrlReader` + other service implementations into service modules
```

```[tasklist]
#### Service Definitions
- [x] [All] Review each service interface individually and list here the desire api-report changes
- [x] [Identity] Deprecate and remove the `IdentityService` definition from `backend-plugin-api`
- [x] [Token Manager] Deprecate and remove the `TokenManagerService` definition from `backend-plugin-api`
- [x] [Url Reader] Evaluate if the `URLReader` is really a core service? Also reconsider if the `Schedule` should continue in the core services list as well
- [ ] https://github.com/backstage/backstage/issues/25134
- [ ] https://github.com/backstage/backstage/issues/25004
- [ ] https://github.com/backstage/backstage/issues/25005
- [ ] https://github.com/backstage/backstage/issues/24869
- [ ] https://github.com/backstage/backstage/issues/25136
```

#### Test Utils

```[tasklist]
### Tasks
- [ ] https://github.com/backstage/backstage/issues/25122
- [ ] https://github.com/backstage/backstage/issues/25128
- [ ] https://github.com/backstage/backstage/issues/25562
- [ ] https://github.com/backstage/backstage/issues/25564
- [ ] https://github.com/backstage/backstage/issues/25563
- [x] The `mockServices` structure should mirror the `coreServices` defined the `backend-plugin-api`
```

### Rollout <a href="#rollout" id="rollout"/>

The following tasks should be done before or in the release pull request:

```[tasklist]
### Tasks
- [ ] https://github.com/backstage/backstage/issues/26000
- [ ] https://github.com/backstage/backstage/issues/26047
- [ ] https://github.com/backstage/backstage/issues/25999
- [ ] https://github.com/backstage/backstage/issues/26046
- [ ] https://github.com/backstage/backstage/issues/26120
- [ ] https://github.com/backstage/backstage/issues/26131
- [ ] https://github.com/backstage/backstage/issues/26121
- [ ] https://github.com/backstage/backstage/issues/26123
- [ ] https://github.com/backstage/backstage/issues/26162
- [ ] https://github.com/backstage/backstage/issues/26160
- [ ] https://github.com/backstage/backstage/issues/26165
- [ ] https://github.com/backstage/backstage/issues/26163
- [ ] https://github.com/backstage/backstage/issues/26169
- [ ] https://github.com/backstage/backstage/issues/26166
- [ ] https://github.com/backstage/backstage/issues/26177
- [ ] https://github.com/backstage/backstage/issues/26164
- [ ] https://github.com/backstage/backstage/issues/26124
- [ ] https://github.com/backstage/backstage/issues/26284
- [ ] https://github.com/backstage/backstage/issues/26283
```

## Nice to Have <a href="#nice-to-have" id="nice-to-have"/>

We've deprioritized the items below. If there's time, they will be done before release.

Disregard:
- [Auth] Prefix `BackstagePrincipalTypes` with "Auth" and investigate the possibility of creating a supertype
- [Auth] The `AuthService.listPublicServiceKeys` seems controversial, possibly create another service to have this method
- [Auth] As a bonus, evaluate adding `expiresAt` to the `AuthService.getPluginRequestToken` return type [Explore if it makes sense]
- [Database] Evaluate removing no longer used entries from `TestDatabaseId`, also consider adding an id for the newest version of each driver (in case we decide to extract these ids to a database package, we should expose them in a test-utils subpath)

Not blocking, apply afterwards:
- [RootHttpRouter] Ensure that the `RootHttpRouterService.use` method holds on plugin request until the backend gets registered

Maybe apply afterwards:
- [Http Auth] Consider allowing people to pass a `path` to the `HttpRouterService.use` method
- [Http Auth] Remove the need to create a router to pass a handler to the `HttpRouterService.use` method. One suggestion is to provide the promise router so people don't forget about using it. Alternatively rename the method to a `getRouter` and guarantee that it always return the same router or modify the `use` method so it could be called multiple times
- [Database] Consider the TestDatabases interface - can it be more explicit and throw errors instead of trying to be forgiving?
- [Documentation] Investigate why in some cases the implementation type is inferred and not the service type, see api-report (low prio, possibly fixed already)

## Out of Scope <a href="#out-of-scope" id="out-of-scope"/>

The items below are related to a new database service exploration, so they are out of scope for this release.

- [Database] Investigate replacing `DatabaseService.migrations` with an `addMigration` method
- [Database] Evaluate creating a new `KnexService` that implements `DatabaseService.getClient` using knex-specific things

**Comment by camilaibs** [comment]
Remaining tasks for the release itself:

To do another small api review
Prepare the main repo for deletion of backend commons
- Update imports (replace backend common imports )
- After that the CI should still green
Prep the 1.0 pr

**Comment by camilaibs** [comment]
We are officially closing this issue after yesterday's [release](https://backstage.io/docs/releases/v1.31.0#backend-system-10-), thanks everyone for helping 🎉
